### PR TITLE
fix(assertions): make is-xml reject non-well-formed XML

### PR DESCRIPTION
**Pull request by i-anubhav-anand**
## Summary

`is-xml` is documented as validating the entire output as XML, but `fast-xml-parser` is a lenient data extractor. It accepted malformed documents such as mismatched tags, duplicate roots, unquoted attributes, invalid comments, trailing text, and malformed `DOCTYPE` declarations.

## Changes

- Validate the full `is-xml` output with `saxes` before using `fast-xml-parser` for required-element lookup.
- Validate the XML 1.0 `DOCTYPE` grammar that `saxes` exposes but does not parse: declaration name/spacing, SYSTEM/PUBLIC literals, public-ID characters, trailing tokens, and fragment-free system identifiers.
- Reject DTD internal subsets because their declarations and replacement text cannot be safely validated without loading DTD content.
- Preserve valid XML edge cases including astral-plane names, BOM/leading whitespace, bare doctypes, and quoted system identifiers containing brackets.
- Keep `contains-xml` intentionally lenient because it extracts XML fragments from otherwise non-XML output.
- Document that `is-xml` requires one well-formed XML document.

## Verification

Validated on head `1e2bd7a09` after merging `origin/main` at `81fe4479e`:

- `npx vitest run test/assertions/xml.test.ts test/assertions/runAssertion.test.ts --sequence.seed=9782` — 209 tests passed.
- 23 valid/invalid external-DOCTYPE forms matched `xmllint --nonet` with no mismatches; the separate no-whitespace case is rejected per the XML 1.0 grammar even though libxml accepts it leniently.
- Real local CLI eval — 4/4 passed: valid external DOCTYPE + required element, leading BOM, missing SYSTEM literal rejected, invalid PUBLIC character rejected.
- `npm run tsc`, `npm run l`, `npm run f`, and `npm run architecture:check` — passed.
- `SKIP_OG_GENERATION=true npm run build` in `site/` — production docs build passed.
- `npm ls saxes --depth=0` — direct dependency resolves to `saxes@6.0.0`.

Reference: [W3C XML 1.0 (Fifth Edition)](https://www.w3.org/TR/xml/).

### CODE REVIEW

**Comment by mldangelo** [comment]
Reviewed and tested end-to-end (real `promptfoo eval`: well-formed XML accepted, malformed rejected; 33 `xml` tests green; `tsc` clean). Security-verified against the real `saxes` parser: XXE (`file:///…`) and billion-laughs both require an internal DTD subset, which the PR rejects wholesale, and `saxes` loads no external resources — no SSRF. Promoting `saxes` to a direct dep with a minimal lockfile change is correct.

**Pushed** a small follow-up: the DTD-internal-subset rejection now reports its clean reason instead of the misleading `XML parsing failed: …` prefix (genuine parse failures keep the prefix), and tightened the test to exact-reason equality. Messaging-only — no input's pass/fail outcome changes (verified directly against `saxes`).

Heads-up (pre-existing, out of scope): a well-formed doc with a leading BOM/whitespace is still rejected by the `startsWith('<')` check with a slightly misleading "missing opening tag" reason. ✅

**Comment by mldangelo** [comment]
Thanks — addressed. The 4 uncovered patch lines codecov flagged are the quote-tracking branch inside `hasDoctypeInternalSubset` (src/assertions/xml.ts:77-83). saxes hands the full DOCTYPE body — including SYSTEM/PUBLIC literals — to the `doctype` handler, so that branch exists to ensure a `[` inside a quoted system/public identifier (e.g. `<!DOCTYPE root SYSTEM "weird[name].dtd">`) isn't misread as the start of an internal subset and rejected. The existing "reject DTD internal subsets" test only used unquoted `[`, so the quote path never ran. I added a focused positive regression test ("should accept external DOCTYPEs whose quoted identifiers contain brackets") asserting both a quoted-bracket SYSTEM and PUBLIC doctype validate as valid XML. That exercises all four lines (verified: each now has coverage hits), suite goes 33 -> 34 green, and tsc is clean. Test-only change, no production behavior touched.


_Pushed in b9c469b8cb._

File: src/assertions/xml.ts [hunk 1]
```diff
@@ -375,6 +396,29 @@ export function validateXml(
...
 ): { isValid: boolean; reason: string } {
+  // XMLParser.parse() is a lenient data extractor, so validate the entire
+  // document with a well-formedness parser before extracting its values.
+  // contains-xml intentionally remains lenient because it accepts fragments
+  // embedded in otherwise non-XML output.
+  let unsupportedReason: string | undefined;
+  try {
+    const parser = new SaxesParser();
+    parser.on('doctype', (doctype) => {
+      if (hasDoctypeInternalSubset(doctype)) {
```
[the remark this answers is not in the document]
Comment (mldangelo-oai) [reply src/assertions/xml.ts:530]: Fixed in 1e2bd7a09. The is-xml path now validates the XML 1.0 DOCTYPE grammar that saxes leaves opaque: required name/whitespace, SYSTEM and PUBLIC literal arity, PUBLIC character set, trailing tokens, and fragment-free system IDs. Added regression cases for the reported examples plus numeric names, junk tokens, invalid public IDs, and missing literals. I also corrected the prior PUBLIC bracket test (brackets are valid in SYSTEM literals, not PubidLiteral) and covered BOM/leading whitespace. Verified with 209 assertion tests, a 23-case xmllint --nonet differential matrix, and a 4/4 real CLI eval.

**Comment by mldangelo-oai** [comment]
@codex review

**Comment by mldangelo-oai** [comment]
@codex review

**Comment by mldangelo-oai** [comment]
@codex review

**Comment by mldangelo-oai** [comment]
@codex review

**Review by mldangelo-oai — APPROVED** [review]
Verified on the final head: focused XML tests, typecheck, docs build, CLI QA, independent review, Codex review, Copilot review, and exact-head CI are green. Ready to merge.

**Comment by mldangelo-oai** [comment]
Thank you, @i-anubhav-anand — this is a careful, well-tested improvement to XML validation. The parser-based approach closes the malformed-XML gap while preserving the supported DTD cases, and the regression coverage made the edge behavior easy to verify. I appreciate the thoroughness here. Merging this now.
